src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java

```diff
@@ -1,43 +1,65 @@
 package com.volasoftware.tinder.services.implementations;
 
+import com.volasoftware.tinder.constants.MailConstant;
+import com.volasoftware.tinder.constants.SecurityConstant;
+import com.volasoftware.tinder.exceptions.EmailAlreadyVerifiedException;
+import com.volasoftware.tinder.exceptions.VerificationTokenExpiredException;
 import com.volasoftware.tinder.models.Account;
 import com.volasoftware.tinder.models.VerificationToken;
 import com.volasoftware.tinder.repositories.VerificationTokenRepository;
-import com.volasoftware.tinder.services.contracts.EmailService;
 import com.volasoftware.tinder.services.contracts.VerificationTokenService;
 import lombok.RequiredArgsConstructor;
-import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
 import java.time.LocalDateTime;
-import java.time.temporal.ChronoUnit;
+import java.util.Optional;
 import java.util.UUID;
 
 @Service
 @RequiredArgsConstructor
 public class VerificationTokenServiceImpl implements VerificationTokenService {
 
-    @Autowired
-    private VerificationTokenRepository verificationTokenRepository;
-
-    @Autowired
-    private EmailService emailService;
+    private final VerificationTokenRepository verificationTokenRepository;
 
     @Override
-    public void generateToken(Account account) {
-
+    public VerificationToken generateToken(Account account) {
         String uuidToken = UUID.randomUUID().toString();
         VerificationToken verificationToken = new VerificationToken(
                 uuidToken,
                 LocalDateTime.now(),
                 LocalDateTime.now(),
-                LocalDateTime.now().plus(2, ChronoUnit.DAYS),
+                LocalDateTime.now().plusDays(SecurityConstant.TOKEN_EXPIRATION_DAYS),
                 account
         );
 
-        verificationTokenRepository.save(verificationToken);
+        return verificationTokenRepository.save(verificationToken);
+    }
+
+    @Override
+    public Account verifyToken(String token) {
+
+        VerificationToken verificationToken = verificationTokenRepository.findByToken(token)
+                .orElseThrow(() -> new VerificationTokenExpiredException(SecurityConstant.TOKEN_EXPIRED));
 
-        emailService.send(account.getEmail(), account.getFirstName(), uuidToken);
+        isEmailAlreadyVerified(verificationToken);
```

Comment:
One idea from me, not saying that must be implemented but for example say that the verification token expires and the user still has not validated his/her email, you want to clear this information by using some kind of schedular, because its obsolete. You can make sure user email has been validated by checking his profile validated flag, and not by checking the token get verified date is null. Because you can make the schedular clear also the tokens for which users have been verified, and not keeping unecessary data in the database. Think about it :)

Reply:
The token is deleted only when it is expired or the user had verified the account by using valid one. If the user uses expired token or clicks again the link in the verification email we can just say that the token is invalid or has been used already. That was my idea.